File: app/src/main/java/com/keepassdroid/crypto/keyDerivation/KdfParameters.java
/*
 * Copyright 2017 Brian Pellin, Jeremy Jamet / Kunzisoft.
 *
 * This file is part of KeePass DX.
 *
 *  KeePass DX is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  KeePass DX is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with KeePass DX.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
package com.keepassdroid.crypto.keyDerivation;

import com.keepassdroid.collections.VariantDictionary;
import com.keepassdroid.stream.LEDataInputStream;
import com.keepassdroid.stream.LEDataOutputStream;
import com.keepassdroid.utils.Types;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.util.UUID;

public class KdfParameters extends VariantDictionary {

    public UUID kdfUUID;

    private static final String ParamUUID = "$UUID";

    public KdfParameters(UUID uuid) {
        kdfUUID = uuid;
    }

    public static KdfParameters deserialize(byte[] data) throws IOException {
        ByteArrayInputStream bis = new ByteArrayInputStream(data);
        LEDataInputStream lis = new LEDataInputStream(bis);

        VariantDictionary d = VariantDictionary.deserialize(lis);
        if (d == null) {
            assert(false);
            return null;
        }

        UUID uuid = Types.bytestoUUID(d.getByteArray(ParamUUID));
        if (uuid == null) {
            assert(false);
            return null;
        }

        KdfParameters kdfP = new KdfParameters(uuid);
        kdfP.copyTo(d);
        return kdfP;

    }

    public static byte[] serialize(KdfParameters kdf) throws IOException {
        ByteArrayOutputStream bos = new ByteArrayOutputStream();
        LEDataOutputStream los = new LEDataOutputStream(bos);

        KdfParameters.serialize(kdf, los);

        return bos.toByteArray();
    }

}

File: app/src/main/java/com/kunzisoft/keepass/database/crypto/kdf/KdfParameters.kt
/*
 * Copyright 2019 Jeremy Jamet / Kunzisoft.
 *
 * This file is part of KeePassDX.
 *
 *  KeePassDX is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  KeePassDX is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with KeePassDX.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
package com.kunzisoft.keepass.database.crypto.kdf

import com.kunzisoft.keepass.utils.bytes16ToUuid
import com.kunzisoft.keepass.utils.uuidTo16Bytes
import com.kunzisoft.keepass.database.crypto.VariantDictionary
import java.io.IOException
import java.util.*

class KdfParameters: VariantDictionary {

    val uuid: UUID

    constructor(uuid: UUID): super() {
        this.uuid = uuid
    }

    constructor(uuid: UUID, d: VariantDictionary): super(d) {
        this.uuid = uuid
    }

    fun setParamUUID() {
        setByteArray(PARAM_UUID, uuidTo16Bytes(uuid))
    }

    companion object {

        private const val PARAM_UUID = "\$UUID"

        @Throws(IOException::class)
        fun deserialize(data: ByteArray): KdfParameters? {
            val dictionary = VariantDictionary.deserialize(data)

            val uuidBytes = dictionary.getByteArray(PARAM_UUID) ?: return null
            val uuid = bytes16ToUuid(uuidBytes)

            return KdfParameters(uuid, dictionary)
        }

        @Throws(IOException::class)
        fun serialize(kdfParameters: KdfParameters): ByteArray {
            return VariantDictionary.serialize(kdfParameters)
        }
    }

}
